### PR TITLE
[OTX] Add SupCon recipes for classification task

### PR DESCRIPTION
**Pull request by arrufat**
This PR contains recipes for training models using a [SupCon](https://arxiv.org/abs/2004.11362)-like style.

These recipes aim to increase the current classification baseline when the data is scarce.

Procedure:
- Train a model with two heads:
  - classification head with standard cross-entropy loss.
  - SupCon head with custom SelfSL loss (currently it uses the [Barlow Twins](https://arxiv.org/abs/2103.03230) loss).
- At inference, the SupCon head is discarded, hence incurring no overhead at runtime.

The recipes include:
- a data pipeline using TwoCrop transform (to generate two views of an image for the SelfSL head)
- a SupConClassifier (in charge of handling the two views of images)
- a SupConClsHead (the one that uses the dual-head approach)
- a SelfSL loss (the Barlow Twins loss)

So far, this method will be used in OTX when specifying the `TrainType` as `SUPERVISEDCONTRASTIVE`.

### CODE REVIEW

File: recipes/stages/classification/supcon.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+optimizer:
+    type: SGD
+    lr: 0.005
+    momentum: 0.9
+    weight_decay: 0.0005
+
+evaluation:
+    metric: ['accuracy', 'class_accuracy']
+
```
Comment (jaegukhyun) [inline recipes/stages/classification/supcon.yaml:21]: Could you explain why task adapt op disappear in this recipe?f

**Review by jaegukhyun — COMMENTED** [review]
Genreally, LGTM. I left one comment

File: recipes/stages/classification/supcon.yaml (hunk 1 above)
Comment (arrufat) [reply recipes/stages/classification/supcon.yaml:21]: Task adapt replaces whichever loss function I specify in the model with CrossEntropyLoss.
And I think it also changed the number of classes or something.

It wasn't working before, but maybe after the refactoring I pushed this morning, it might work.

I will check and revert it if it works.

Comment (arrufat) [reply recipes/stages/classification/supcon.yaml:21]: Yes, I've run a couple of tests, and it works now with `task_adapt` after the refactoring. Thank you!

File: mpa/modules/models/heads/supcon_cls_head.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    def __init__(self, num_classes: int, in_channels: int, aux_mlp, loss, aux_loss, topk=(1,), init_cfg=None, **kwargs):
+        if in_channels <= 0:
+            raise ValueError(f"in_channels={in_channels} must be a positive integer")
+        if num_classes <= 0:
+            raise ValueError("at least one class must be exist num_classes.")
+
+        if isinstance(topk, int):
+            topk = (topk,)
+        for _topk in topk:
+            assert _topk > 0, "Top-k should be larger than 0"
```
Comment (jaegukhyun) [inline mpa/modules/models/heads/supcon_cls_head.py:34]: Is it okay if _topk is larger than num_classes?

**Review by jaegukhyun — COMMENTED** [review]
I left one comment, and I have a question. Where is the part for dividing learning rate by 2? I heard that is necessary because of changes in batch size

File: mpa/modules/models/heads/supcon_cls_head.py (hunk 2 above)
Comment (arrufat) [reply mpa/modules/models/heads/supcon_cls_head.py:34]: Fixed!

**Comment by arrufat** [comment]
> I left one comment, and I have a question. Where is the part for dividing learning rate by 2? I heard that is necessary because of changes in batch size

As discussed with Wonju, for now, this logic is not implemented, we thought it would be better to do it manually and, once it because the default method for small datasets, add the logic in the code. But it's up for discussion.

**Review by jaegukhyun — APPROVED** [review]
LGTM, @harimkang, @supersoob, @JihwanEom anyone, could you check this PR once?

**Review by harimkang — APPROVED** [review]
If this works well, there seems to be no structural problem. LGTM.
